src/frontend/src/expr/table_function.rs

```diff
@@ -226,6 +229,127 @@ impl TableFunction {
         })
     }
 
+    pub fn new_postgres_query(args: Vec<ExprImpl>) -> RwResult<Self> {
+        let args = {
+            if args.len() != 6 {
+                return Err(BindError("postgres_query function only accepts 6 arguments: postgres_query(hostname varchar, port varchar, username varchar, password varchar, database_name varchar, postgres_query varchar)".to_string()).into());
+            }
+            let mut cast_args = Vec::with_capacity(6);
+            for arg in args {
+                let arg = arg.cast_implicit(DataType::Varchar)?;
+                cast_args.push(arg);
+            }
+            cast_args
+        };
+        let evaled_args = {
+            let mut evaled_args: Vec<String> = Vec::with_capacity(6);
+            for arg in &args {
+                match arg.try_fold_const() {
+                    Some(Ok(value)) => {
+                        let Some(scalar) = value else {
+                            return Err(BindError(
+                                "postgres_query function does not accept null arguments"
+                                    .to_string(),
+                            )
+                            .into());
+                        };
+                        evaled_args.push(scalar.into_utf8().into());
+                    }
+                    Some(Err(err)) => {
+                        return Err(err);
+                    }
+                    None => {
+                        return Err(BindError(
+                            "postgres_query function only accepts constant arguments".to_string(),
+                        )
+                        .into());
+                    }
+                }
+            }
+            evaled_args
+        };
+
+        #[cfg(madsim)]
+        {
+            return Err(crate::error::ErrorCode::BindError(
+                "postgres_query can't be used in the madsim mode".to_string(),
+            )
+            .into());
+        }
+
+        #[cfg(not(madsim))]
+        {
+            let schema = tokio::task::block_in_place(|| {
+                RUNTIME.block_on(async {
```

Comment:
May note that this still blocks the current thread, meaning that the concurrency of binding Postgres TVF cannot exceed the `TOKIO_WORKER_THREADS`. As we're introducing more in the future, we may have to consider refactoring this part. Like

- make binder async (can be abused)
- resolve necessary info ahead of time

Reply:
> resolve necessary info ahead of time

I was thinking of the latter approach. Something like splitting the binder into 2 phases, first is async phase which can resolve bindings like this. Second is the current binder, just a pure function.

So it will be something like:

```
async fn bind(&self, ast) {
  let pure_ast = self.bind_async(ast);
  let ast = self.bind_sync(pure_ast);
}

async fn bind_async() {...}
fn bind_sync() {...}
```

Reply:
Just mentioning a third option. I do not prefer it but it may inspire an improved fourth option:
* Let the user specify return schema and defer the async validation (instead of `ahead of time`)

The following syntax exists in PostgreSQL but we have not supported it yet:
```
select * from json_to_recordset('[{"a":1,"b":"foo"}, {"a":"2","c":"bar"}]') as x(a int, b text)
-- Similarly: select * from postgres_query(...) as x(a int, b text)
```

Reply:
> * make binder async (can be abused)

Historical context on binder not being async, although most of the discussion happened in slack private channel:
https://github.com/risingwavelabs/risingwave/pull/173#issuecomment-1034558497

Reply:
So far the simplest way, while preserving good ux, seems to be just making the binder async.

The other approaches in this thread will either make maintainence more cumbersome (extra IR in binder), or lead to complexity in ux.

Reply:
To clarify making binder `async` may not be easy. We can discuss the topic separately.